### PR TITLE
dotCMS/core#16054 Add class to rows and columns in the layout editor

### PR DESCRIPTION
**Pull request by freddyucv**

### CODE REVIEW

File: src/app/portlets/dot-edit-page/layout/components/dot-edit-layout-grid/dot-edit-layout-grid.component.ts [hunk 1]
```diff
@@ -35,8 +31,13 @@ export class DotEditLayoutGridComponent implements OnInit, ControlValueAccessor
     @ViewChild(NgGrid)
     ngGrid: NgGrid;
 
+    @ViewChild('classToAddInput')
+    classToAddInput: ElementRef;
```
Comment (freddyucv) [inline src/app/portlets/dot-edit-page/layout/components/dot-edit-layout-grid/dot-edit-layout-grid.component.ts]: here I want to do in the html something like:

<input pInputText #classToAddInput type="text" [(value)]=''classToAdd"/>

and here something like:

classToAdd: string;

I don't know why but not work in that way,  but I have to try it again.

File: src/app/portlets/dot-edit-page/shared/services/dot-edit-layout.service.ts [hunk 2]
```diff
@@ -70,8 +66,8 @@ export class DotEditLayoutService {
      * @param DotLayoutBody grid
      * @returns DotLayoutBody
      */
-    cleanupDotLayoutBody(grid: DotLayoutBody): DotLayoutBody {
-        return this.getDotLayoutBody(this.getDotLayoutGridBox(grid));
+    cleanupDotLayoutBody(layoutBody: DotLayoutBody): DotLayoutBody {
+        return this.getDotLayoutBody(this.getDotLayoutGridBox(layoutBody));
```
Comment (freddyucv) [inline src/app/portlets/dot-edit-page/shared/services/dot-edit-layout.service.ts]: I think this method is unnecessary, but I am not sure yet

File: src/app/portlets/dot-edit-page/layout/components/dot-edit-layout-grid/dot-edit-layout-grid.component.ts [hunk 3]
```diff
@@ -188,70 +188,70 @@ export class DotEditLayoutGridComponent implements OnInit, ControlValueAccessor
         }
     }
 
-    private setGridValue(): void {
-        this.grid = this.isHaveRows()
-            ? this.dotEditLayoutService.getDotLayoutGridBox(this.value)
-            : [...DOT_LAYOUT_DEFAULT_GRID];
+    addColumnClass(index: number): void {
+        this.addClass(
+            () => this.grid[index].config.payload.styleClass || '',
```
Comment (freddyucv) [inline src/app/portlets/dot-edit-page/layout/components/dot-edit-layout-grid/dot-edit-layout-grid.component.ts]: maybe it should be 'null', not ''

File: src/app/portlets/dot-edit-page/shared/models/dot-layout-grid.model.ts [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+    sizex: 3,
+    maxCols: 12,
+    maxRows: 1
+};
+
+/**
+ * TODO
+ *
+ */
+export class DotLayoutGrid {
```
Comment (freddyucv) [inline src/app/portlets/dot-edit-page/shared/models/dot-layout-grid.model.ts]: Java doc

File: src/app/portlets/dot-edit-page/layout/components/dot-edit-layout-grid/dot-edit-layout-grid.component.ts [hunk 5]
```diff
@@ -188,70 +188,75 @@ export class DotEditLayoutGridComponent implements OnInit, ControlValueAccessor
         }
     }
 
-    private setGridValue(): void {
-        this.grid = this.isHaveRows()
-            ? this.dotEditLayoutService.getDotLayoutGridBox(this.value)
-            : [...DOT_LAYOUT_DEFAULT_GRID];
-    }
+    addColumnClass(index: number): void {
```
Comment (freddyucv) [inline src/app/portlets/dot-edit-page/layout/components/dot-edit-layout-grid/dot-edit-layout-grid.component.ts]: JavaDoc

File: src/app/portlets/dot-edit-page/layout/components/dot-edit-layout-grid/dot-edit-layout-grid.component.ts [hunk 6]
```diff
@@ -188,70 +188,75 @@ export class DotEditLayoutGridComponent implements OnInit, ControlValueAccessor
...
+                } else {
+                    this.grid.boxes[index].config.payload.styleClass = value;
+                }
+            }
+        );
     }
 
-    private setConfigOfNewContainer(): any {
-        const newRow: any = Object.assign({}, DOT_LAYOUT_GRID_NEW_ROW_TEMPLATE);
+    addRowClass(index: number): void {
```
Comment (freddyucv) [inline src/app/portlets/dot-edit-page/layout/components/dot-edit-layout-grid/dot-edit-layout-grid.component.ts]: Java Doc

File: src/app/portlets/dot-edit-page/layout/components/dot-edit-layout-grid/dot-edit-layout-grid.component.ts [hunk 7]
```diff
@@ -188,70 +188,78 @@ export class DotEditLayoutGridComponent implements OnInit, ControlValueAccessor
...
-    private setConfigOfNewContainer(): any {
-        const newRow: any = Object.assign({}, DOT_LAYOUT_GRID_NEW_ROW_TEMPLATE);
+    /**
+     * Add style class to a row
+     * @param index row index
+     */
+    addRowClass(index: number): void {
+        this.addClass(
+            () => this.grid.getRowClass(index) || '',
+            () => this.grid.setRowClass(this.classToAddInput.nativeElement.value, index)
```
Comment (fmontes) [inline src/app/portlets/dot-edit-page/layout/components/dot-edit-layout-grid/dot-edit-layout-grid.component.ts]: `this.classToAddInput.nativeElement.value` should be a ng form

File: src/app/portlets/dot-edit-page/layout/components/dot-edit-layout-grid/dot-edit-layout-grid.component.ts [hunk 8]
```diff
@@ -188,70 +188,78 @@ export class DotEditLayoutGridComponent implements OnInit, ControlValueAccessor
...
+        this.classToAddInput.nativeElement.value = getterFunc.bind(this)();
+        this.showAddClassDialog = true;
+    }
 
-            busyColumns += lastContainer.config.col - 1 + lastContainer.config.sizex;
+    private setGridValue(): void {
+        this.grid = this.isHaveRows()
+            ? this.dotEditLayoutService.getDotLayoutGridBox(this.value)
+            : DotLayoutGrid.getDefaultGrid();
+        console.log(this.grid);
```
Comment (fmontes) [inline src/app/portlets/dot-edit-page/layout/components/dot-edit-layout-grid/dot-edit-layout-grid.component.ts]: remove

File: src/app/portlets/dot-edit-page/layout/components/dot-edit-layout-grid/dot-edit-layout-grid.component.ts [hunk 9]
```diff
@@ -188,70 +188,78 @@ export class DotEditLayoutGridComponent implements OnInit, ControlValueAccessor
...
     }
 
-    private setConfigOfNewContainer(): any {
-        const newRow: any = Object.assign({}, DOT_LAYOUT_GRID_NEW_ROW_TEMPLATE);
+    /**
+     * Add style class to a row
+     * @param index row index
+     */
+    addRowClass(index: number): void {
+        this.addClass(
```
Comment (fmontes) [inline src/app/portlets/dot-edit-page/layout/components/dot-edit-layout-grid/dot-edit-layout-grid.component.ts]: This whole code to add class looks very confusing.

File: src/app/portlets/dot-edit-page/shared/models/dot-layout-grid.model.spec.ts [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+        expect(DotLayoutGrid.getDefaultConfig()).toEqual({
+            fixed: true,
+            sizex: 3,
+            maxCols: 12,
+            maxRows: 1
+        });
+    });
+
+    it('should get default grid layout object', () => {
+        console.log('DotLayoutGrid.getDefaultGrid()', DotLayoutGrid.getDefaultGrid());
```
Comment (fmontes) [inline src/app/portlets/dot-edit-page/shared/models/dot-layout-grid.model.spec.ts]: remove

File: src/app/portlets/dot-edit-page/layout/components/dot-edit-layout-grid/dot-edit-layout-grid.component.html [hunk 11]
```diff
@@ -1,8 +1,42 @@
...
+        >
         </dot-container-selector>
+        <div class="box__actions">
+            <dot-icon-button icon="more_horiz" (click)="addColumnClass(i)"></dot-icon-button>
+            <dot-icon-button icon="close" (click)="onRemoveContainer(i)"></dot-icon-button>
+        </div>
     </div>
 </div>
+
+<dot-dialog
```
Comment (fmontes) [inline src/app/portlets/dot-edit-page/layout/components/dot-edit-layout-grid/dot-edit-layout-grid.component.html]: We need to disable and enable buttons when form is valid, invalid, etc.. also focus in the field.

**Pull request merged**